src/app/icd/client/ICDClientInfoPersistentStorage.h

```diff
@@ -0,0 +1,95 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include "ICDClientInfo.h"
+#include "ICDStorageKeyDelegate.h"
+#include <crypto/CHIPCryptoPAL.h>
+#include <lib/core/CHIPConfig.h>
+#include <lib/core/CHIPPersistentStorageDelegate.h>
+#include <lib/core/DataModelTypes.h>
+#include <lib/core/ScopedNodeId.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/CommonIterator.h>
+#include <lib/support/StorageKeyName.h>
+#include <stddef.h>
+
+namespace chip {
+namespace app {
+
+/**
+ * The ICDClientInfoPersistentStorage class is an abstract interface that defines the operations
+ * for storing, retrieving and deleting ICD client information in persistent storage.
+ */
+class ICDClientInfoPersistentStorage
+{
+public:
+    using ICDClientInfoIterator = CommonIterator<ICDClientInfo>;
+
+    virtual ~ICDClientInfoPersistentStorage() = default;
+
+    /**
+     * Iterate through persisted ICD Client Info
+     *
+     * @return A valid iterator on success. Use CommonIterator accessor to retrieve ICDClientInfo
+     */
+    virtual ICDClientInfoIterator * IterateICDClientInfo() = 0;
+
+    /**
+     * Called during ICD device registration in commissioning, sdk provides the clientInfo and application
+     * provides raw key data, the shared key handle in clientInfo is updated based upon raw key data
+     *
+     * @param[inout] aICDClientInfo the ICD Client information to be updated with keyData and be saved
+     * @param[in] aKeyData raw key data provided by application
+     */
+    virtual CHIP_ERROR SetKey(ICDClientInfo & aClientInfo, const ByteSpan aKeyData) = 0;
```

Comment:
The `aClientInfo` format for arguments is no longer used in any new code, and this is a brand new module. The code is written now, but consider avoiding this format in newer code.